File: pi/adapter.ts
/**
 * Frontier model adapter — standalone, testable core.
 *
 * No pi dependencies. Pure functions using only fetch and process.env.
 * Used by both the pi extension and any future integrations.
 */

// ── Configuration ──────────────────────────────────────────────────────────

import { ProxyAgent } from "undici";

export const DEFAULT_SYSTEM_PROMPT =
  "You are being consulted as a frontier advisory model by a local AI system " +
  "that handles most tasks independently. You are called only when the local " +
  "model has determined it needs capabilities beyond its own. " +
  "Be direct and substantive. " +
  "Do not repeat the question back. Do not pad with caveats. " +
  "The local model is technically competent -- treat it as a peer.";

export const MAX_TOKENS = 4096;

export const MODEL_PREFERENCE = [
  ["anthropic", "claude-opus-4-7"] as const,
  ["openai", "gpt-4.1"] as const,
] as const;

export type ProviderName = (typeof MODEL_PREFERENCE)[number][0];

export const PROVIDER_CONFIG: Record<ProviderName, {
  envKey: string;
  envBaseUrl: string;
  defaultBaseUrl: string;
}> = {
  anthropic: {
    envKey: "ANTHROPIC_API_KEY",
    envBaseUrl: "ANTHROPIC_BASE_URL",
    defaultBaseUrl: "https://api.anthropic.com",
  },
  openai: {
    envKey: "OPENAI_API_KEY",
    envBaseUrl: "OPENAI_BASE_URL",
    defaultBaseUrl: "https://api.openai.com",
  },
};

// ── Credential resolution ─────────────────────────────────────────────────

export type ApiKeyResolver = (provider: ProviderName) => { apiKey: string; baseUrl: string } | undefined | Promise<{ apiKey: string; baseUrl: string } | undefined>;

export function envKeyResolver(): ApiKeyResolver {
  return (provider: ProviderName) => {
    const cfg = PROVIDER_CONFIG[provider];
    const apiKey = process.env[cfg.envKey];
    if (!apiKey) return undefined;
    return {
      apiKey,
      baseUrl: process.env[cfg.envBaseUrl] ?? cfg.defaultBaseUrl,
    };
  };
}

function isPromise<T>(value: T | Promise<T>): value is Promise<T> {
  return typeof (value as Promise<T>)?.then === "function";
}

export function getProviderCredentials(provider: ProviderName, resolver?: ApiKeyResolver): { apiKey: string; baseUrl: string } | undefined {
  const resolve = resolver ?? envKeyResolver();
  const result = resolve(provider);
  return isPromise(result) ? undefined : result;
}

export async function getProviderCredentialsAsync(provider: ProviderName, resolver?: ApiKeyResolver): Promise<{ apiKey: string; baseUrl: string } | undefined> {
  const resolve = resolver ?? envKeyResolver();
  return resolve(provider);
}

// ── Proxy support for fetch ────────────────────────────────────────────────

let _proxyDispatcher: ReturnType<typeof ProxyAgent> | undefined;

function getDispatcher(): typeof fetch | undefined {
  // Respect proxy env vars (http_proxy, https_proxy, HTTPS_PROXY, HTTP_PROXY)
  // Undici's fetch in Docker doesn't auto-respect these, so we wire it explicitly.
  const proxyUrl =
    process.env.https_proxy ||
    process.env.HTTPS_PROXY ||
    process.env.http_proxy ||
    process.env.HTTP_PROXY;

  if (!proxyUrl) return undefined;

  // Avoid creating a new dispatcher on every call — cache it
  if (!_proxyDispatcher) {
    try {
      _proxyDispatcher = new ProxyAgent({ uri: proxyUrl });
    } catch {
      // ProxyAgent construction can fail in some environments; fall through to plain fetch
      _proxyDispatcher = undefined;
    }
  }

  return _proxyDispatcher;
}

// ── HTTP clients ───────────────────────────────────────────────────────────

export interface ConsultResult {
  text: string;
  inputTokens: number;
  outputTokens: number;
}

export async function callAnthropic(
  model: string,
  question: string,
  context: string,
  maxTokens: number,
  systemPrompt: string,
  creds: { apiKey: string; baseUrl: string },
): Promise<ConsultResult> {
  const headers: Record<string, string> = {
    "x-api-key": creds.apiKey,
    "anthropic-version": "2023-06-01",
    "content-type": "application/json",
  };

  const userContent = context
    ? `<advisory_context>\n${context}\n</advisory_context>\n\n<advisory_question>\n${question}\n</advisory_question>`
    : question;

  const body = JSON.stringify({
    model,
    max_tokens: maxTokens,
    system: systemPrompt,
    messages: [{ role: "user", content: userContent }],
  });

  const dispatcher = getDispatcher();
  const resp = await fetch(`${creds.baseUrl}/v1/messages`, {
    method: "POST",
    headers,
    body,
    dispatcher,
  });

  if (!resp.ok) {
    const bodyText = await resp.text();
    throw new Error(`Anthropic ${resp.status}: ${bodyText}`);
  }

  const data = await resp.json() as {
    content: Array<{ type: string; text?: string }>;
    usage: { input_tokens: number; output_tokens: number };
  };

  const text = data.content.filter((b) => b.type === "text").map((b) => b.text ?? "").join("");

  return {
    text,
    inputTokens: data.usage.input_tokens,
    outputTokens: data.usage.output_tokens,
  };
}

export async function callOpenAI(
  model: string,
  question: string,
  context: string,
  maxTokens: number,
  systemPrompt: string,
  creds: { apiKey: string; baseUrl: string },
): Promise<ConsultResult> {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${creds.apiKey}`,
    "Content-Type": "application/json",
  };

  const msgs = [{ role: "system" as const, content: systemPrompt }];
  const userContent = context ? `Context:\n${context}\n\nQuestion:\n${question}` : question;
  msgs.push({ role: "user" as const, content: userContent });

  const body = JSON.stringify({ model, max_tokens: maxTokens, messages: msgs });

  const dispatcher = getDispatcher();
  const resp = await fetch(`${creds.baseUrl}/v1/chat/completions`, {
    method: "POST",
    headers,
    body,
    dispatcher,
  });

  if (!resp.ok) {
    const bodyText = await resp.text();
    throw new Error(`OpenAI ${resp.status}: ${bodyText}`);
  }

  const data = await resp.json() as {
    choices: Array<{ message: { content: string } }>;
    usage: { prompt_tokens: number; completion_tokens: number };
  };

  return {
    text: data.choices[0]?.message.content ?? "",
    inputTokens: data.usage.prompt_tokens,
    outputTokens: data.usage.completion_tokens,
  };
}

// ── Consult core logic ─────────────────────────────────────────────────────

export interface ConsultOptions {
  question: string;
  context?: string;
  systemPrompt?: string;
  signal?: AbortSignal;
  /** Custom key resolver (e.g. from pi's AuthStorage). Defaults to process.env. */
  apiKeyResolver?: ApiKeyResolver;
}

export interface ConsultResponse {
  response: string;
  provider: string;
  model: string;
  inputTokens: number;
  outputTokens: number;
  latencyMs: number;
}

export async function consult({ question, context = "", systemPrompt, apiKeyResolver }: ConsultOptions): Promise<ConsultResponse> {
  const sysPrompt = systemPrompt ?? DEFAULT_SYSTEM_PROMPT;
  let lastError: unknown;

  const start = performance.now();

  for (const [provider, modelId] of MODEL_PREFERENCE) {
    const creds = await getProviderCredentialsAsync(provider as ProviderName, apiKeyResolver);
    if (!creds) continue;

    try {
      const result = await (provider === "anthropic"
        ? callAnthropic(modelId, question, context, MAX_TOKENS, sysPrompt, creds)
        : callOpenAI(modelId, question, context, MAX_TOKENS, sysPrompt, creds));

      const latencyMs = Math.round(performance.now() - start);

      return {
        response: result.text,
        provider,
        model: modelId,
        inputTokens: result.inputTokens,
        outputTokens: result.outputTokens,
        latencyMs,
      };
    } catch (e) {
      console.error(`[frontier-advisor] Provider ${provider}/${modelId} failed:`, e);
      lastError = e;
    }
  }

  if (lastError) {
    throw new Error(`All configured providers failed. Last error: ${lastError}`);
  }
  throw new Error(
    "No API key configured for any provider. Set ANTHROPIC_API_KEY and/or OPENAI_API_KEY in the environment.",
  );
}

File: mcp/src/frontier_advisor/adapter.py
"""Frontier model adapter with provider fallback.

Credentials from environment variables:
  ANTHROPIC_API_KEY  — Anthropic API key
  OPENAI_API_KEY     — OpenAI API key

Base URLs default to the public APIs. Override with:
  ANTHROPIC_BASE_URL — default: https://api.anthropic.com
  OPENAI_BASE_URL    — default: https://api.openai.com

With mcp-vault, these become vault: references in mcp.json env block,
resolved transparently at process spawn.
"""

import logging
import os
import time

import httpx

logger = logging.getLogger(__name__)

DEFAULT_SYSTEM_PROMPT = (
    "You are being consulted as a frontier advisory model by a local AI system "
    "that handles most tasks independently. You are called only when the local "
    "model has determined it needs capabilities beyond its own. "
    "Be direct, substantive, and efficient with tokens. "
    "Do not repeat the question back. Do not pad with caveats. "
    "The local model is technically competent -- treat it as a peer."
)

MAX_TOKENS = 4096

MODEL_PREFERENCE = [
    ("anthropic", "claude-opus-4-7"),
    ("openai", "gpt-4.1"),
]

PROVIDER_CONFIG = {
    "anthropic": {
        "env_key": "ANTHROPIC_API_KEY",
        "env_base_url": "ANTHROPIC_BASE_URL",
        "default_base_url": "https://api.anthropic.com",
    },
    "openai": {
        "env_key": "OPENAI_API_KEY",
        "env_base_url": "OPENAI_BASE_URL",
        "default_base_url": "https://api.openai.com",
    },
}


class FrontierAdapter:
    def __init__(self):
        self._client = httpx.AsyncClient(timeout=120.0)

    def _get_provider_config(self, provider: str) -> dict | None:
        """Read provider credentials from environment."""
        cfg = PROVIDER_CONFIG.get(provider)
        if not cfg:
            return None
        api_key = os.environ.get(cfg["env_key"])
        if not api_key:
            return None
        return {
            "api_key": api_key,
            "base_url": os.environ.get(cfg["env_base_url"], cfg["default_base_url"]),
        }

    async def consult(
        self,
        question: str,
        context: str = "",
        system_prompt: str | None = None,
    ) -> dict:
        """Send question to frontier model with provider fallback.

        Tries each model in preference order until one succeeds.
        Returns dict with response, provider, model, token counts, latency.
        Raises RuntimeError if all providers fail.
        """
        sys_prompt = system_prompt or DEFAULT_SYSTEM_PROMPT
        last_error = None

        for provider, model_id in MODEL_PREFERENCE:
            creds = self._get_provider_config(provider)
            if not creds:
                continue
            try:
                start = time.monotonic()
                result = await self._call(
                    provider, model_id, question, context,
                    MAX_TOKENS, sys_prompt, creds,
                )
                latency = int((time.monotonic() - start) * 1000)
                return {
                    "response": result["text"],
                    "provider": provider,
                    "model": model_id,
                    "input_tokens": result["input_tokens"],
                    "output_tokens": result["output_tokens"],
                    "latency_ms": latency,
                }
            except Exception as e:
                logger.warning("Provider %s/%s failed: %s", provider, model_id, e)
                last_error = e

        raise RuntimeError(
            f"No provider available. Last error: {last_error}"
        )

    async def _call(self, provider, model, q, ctx, max_tok, sys_prompt, creds):
        if provider == "anthropic":
            return await self._anthropic(model, q, ctx, max_tok, sys_prompt, creds)
        return await self._openai(model, q, ctx, max_tok, sys_prompt, creds)

    async def _anthropic(self, model, q, ctx, max_tok, sys_prompt, creds):
        headers = {
            "x-api-key": creds["api_key"],
            "anthropic-version": "2023-06-01",
            "content-type": "application/json",
        }
        user_content = q
        if ctx:
            user_content = (
                f"<advisory_context>\n{ctx}\n</advisory_context>\n\n"
                f"<advisory_question>\n{q}\n</advisory_question>"
            )
        body = {
            "model": model,
            "max_tokens": max_tok,
            "system": sys_prompt,
            "messages": [{"role": "user", "content": user_content}],
        }
        resp = await self._client.post(
            f"{creds['base_url']}/v1/messages", headers=headers, json=body,
        )
        resp.raise_for_status()
        data = resp.json()
        text = "".join(b["text"] for b in data["content"] if b["type"] == "text")
        return {
            "text": text,
            "input_tokens": data["usage"]["input_tokens"],
            "output_tokens": data["usage"]["output_tokens"],
        }

    async def _openai(self, model, q, ctx, max_tok, sys_prompt, creds):
        headers = {
            "Authorization": f"Bearer {creds['api_key']}",
            "Content-Type": "application/json",
        }
        msgs = [{"role": "system", "content": sys_prompt}]
        user_content = f"Context:\n{ctx}\n\nQuestion:\n{q}" if ctx else q
        msgs.append({"role": "user", "content": user_content})
        body = {
            "model": model,
            "max_tokens": max_tok,
            "messages": msgs,
        }
        resp = await self._client.post(
            f"{creds['base_url']}/v1/chat/completions", headers=headers, json=body,
        )
        resp.raise_for_status()
        data = resp.json()
        return {
            "text": data["choices"][0]["message"]["content"],
            "input_tokens": data["usage"]["prompt_tokens"],
            "output_tokens": data["usage"]["completion_tokens"],
        }
